<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Projet-4-:-Analysez-les-ventes-de-votre-entreprise" data-toc-modified-id="Projet-4-:-Analysez-les-ventes-de-votre-entreprise-1">Projet 4 : Analysez les ventes de votre entreprise</a></span><ul class="toc-item"><li><span><a href="#Analyse-globale-des-données" data-toc-modified-id="Analyse-globale-des-données-1.1">Analyse globale des données</a></span></li><li><span><a href="#1.-df_customers" data-toc-modified-id="1.-df_customers-1.2">1. df_customers</a></span><ul class="toc-item"><li><span><a href="#Apperçu-de-la-population" data-toc-modified-id="Apperçu-de-la-population-1.2.1">Apperçu de la population</a></span></li><li><span><a href="#Déscription-générale-des-données-quantitatifs" data-toc-modified-id="Déscription-générale-des-données-quantitatifs-1.2.2">Déscription générale des données quantitatifs</a></span></li><li><span><a href="#Vérification-des-valeurs-nulles" data-toc-modified-id="Vérification-des-valeurs-nulles-1.2.3">Vérification des valeurs nulles</a></span></li><li><span><a href="#Vérification-des-erreurs-lexicales" data-toc-modified-id="Vérification-des-erreurs-lexicales-1.2.4">Vérification des erreurs lexicales</a></span></li><li><span><a href="#Vérification-des-individus-dupliqués" data-toc-modified-id="Vérification-des-individus-dupliqués-1.2.5">Vérification des individus dupliqués</a></span></li></ul></li><li><span><a href="#2.-df_products" data-toc-modified-id="2.-df_products-1.3">2. df_products</a></span><ul class="toc-item"><li><span><a href="#Apperçu-de-la-population" data-toc-modified-id="Apperçu-de-la-population-1.3.1">Apperçu de la population</a></span></li><li><span><a href="#Déscription-générale-des-données-quantitatifs" data-toc-modified-id="Déscription-générale-des-données-quantitatifs-1.3.2">Déscription générale des données quantitatifs</a></span></li><li><span><a href="#Détail-du-produit-avec-prix-négatif" data-toc-modified-id="Détail-du-produit-avec-prix-négatif-1.3.3">Détail du produit avec prix négatif</a></span></li><li><span><a href="#Vérification-des-valeurs-nulles" data-toc-modified-id="Vérification-des-valeurs-nulles-1.3.4">Vérification des valeurs nulles</a></span></li><li><span><a href="#Vérification-des-erreurs-lexicales" data-toc-modified-id="Vérification-des-erreurs-lexicales-1.3.5">Vérification des erreurs lexicales</a></span></li></ul></li><li><span><a href="#3.-df_transactions" data-toc-modified-id="3.-df_transactions-1.4">3. df_transactions</a></span><ul class="toc-item"><li><span><a href="#Apperçu-de-la-population" data-toc-modified-id="Apperçu-de-la-population-1.4.1">Apperçu de la population</a></span></li><li><span><a href="#Déscription-générale-des-données-quantitatifs" data-toc-modified-id="Déscription-générale-des-données-quantitatifs-1.4.2">Déscription générale des données quantitatifs</a></span></li><li><span><a href="#Détail-des-dates-'test'" data-toc-modified-id="Détail-des-dates-'test'-1.4.3">Détail des dates 'test'</a></span></li><li><span><a href="#Clients-crées-pour-les-dates-'test'" data-toc-modified-id="Clients-crées-pour-les-dates-'test'-1.4.4">Clients crées pour les dates 'test'</a></span></li><li><span><a href="#Vérification-des-valeurs-manquantes" data-toc-modified-id="Vérification-des-valeurs-manquantes-1.4.5">Vérification des valeurs manquantes</a></span></li></ul></li><li><span><a href="#4.-Relation-entre-les-3-DF" data-toc-modified-id="4.-Relation-entre-les-3-DF-1.5">4. Relation entre les 3 DF</a></span><ul class="toc-item"><li><span><a href="#Vérifier-que-tous-les-'client_id'-de-transaction-appartiennent-à-ceux-de-customers" data-toc-modified-id="Vérifier-que-tous-les-'client_id'-de-transaction-appartiennent-à-ceux-de-customers-1.5.1">Vérifier que tous les 'client_id' de transaction appartiennent à ceux de customers</a></span></li><li><span><a href="#Vérifier-que-tous-les-'id_product'-de-transaction-appartiennent-à-ceux-de-products" data-toc-modified-id="Vérifier-que-tous-les-'id_product'-de-transaction-appartiennent-à-ceux-de-products-1.5.2">Vérifier que tous les 'id_product' de transaction appartiennent à ceux de products</a></span></li></ul></li><li><span><a href="#Nettoyage-des-données" data-toc-modified-id="Nettoyage-des-données-1.6">Nettoyage des données</a></span><ul class="toc-item"><li><span><a href="#Supprimer-les-dates-&quot;test&quot;" data-toc-modified-id="Supprimer-les-dates-&quot;test&quot;-1.6.1">Supprimer les dates "test"</a></span></li><li><span><a href="#Vérifier-si-l'un-des-idendifiants-assiciés-aux-test-existe-dans-les-autres-bd" data-toc-modified-id="Vérifier-si-l'un-des-idendifiants-assiciés-aux-test-existe-dans-les-autres-bd-1.6.2">Vérifier si l'un des idendifiants assiciés aux test existe dans les autres bd</a></span></li><li><span><a href="#Suppression-des-clients-et-produits-relatifs-aux-dates-test" data-toc-modified-id="Suppression-des-clients-et-produits-relatifs-aux-dates-test-1.6.3">Suppression des clients et produits relatifs aux dates test</a></span></li><li><span><a href="#Rajout-du-produit--0_2245" data-toc-modified-id="Rajout-du-produit--0_2245-1.6.4">Rajout du produit  0_2245</a></span></li></ul></li></ul></li><li><span><a href="#Jointure-des-données" data-toc-modified-id="Jointure-des-données-2">Jointure des données</a></span><ul class="toc-item"><li><ul class="toc-item"><li><span><a href="#Formatage-des-dates" data-toc-modified-id="Formatage-des-dates-2.0.1">Formatage des dates</a></span></li></ul></li><li><span><a href="#Rajout-des-colonnes" data-toc-modified-id="Rajout-des-colonnes-2.1">Rajout des colonnes</a></span><ul class="toc-item"><li><span><a href="#age" data-toc-modified-id="age-2.1.1">age</a></span></li><li><span><a href="#age_range-et--age_range_mean" data-toc-modified-id="age_range-et--age_range_mean-2.1.2">age_range et  age_range_mean</a></span></li><li><span><a href="#month" data-toc-modified-id="month-2.1.3">month</a></span></li><li><span><a href="#price_purchase" data-toc-modified-id="price_purchase-2.1.4">price_purchase</a></span></li><li><span><a href="#price_range" data-toc-modified-id="price_range-2.1.5">price_range</a></span></li><li><span><a href="#number_purchase_month_age,-number_purchase_age,-purchase_frequency_month-et-mean_purchase_age" data-toc-modified-id="number_purchase_month_age,-number_purchase_age,-purchase_frequency_month-et-mean_purchase_age-2.1.6">number_purchase_month_age, number_purchase_age, purchase_frequency_month et mean_purchase_age</a></span></li><li><span><a href="#number_purchase" data-toc-modified-id="number_purchase-2.1.7">number_purchase</a></span></li><li><span><a href="#basket_size" data-toc-modified-id="basket_size-2.1.8">basket_size</a></span></li><li><span><a href="#basket_size" data-toc-modified-id="basket_size-2.1.9">basket_size</a></span></li></ul></li><li><span><a href="#Suppression-du-mois-d'octobre-pour-la-catégorie-1" data-toc-modified-id="Suppression-du-mois-d'octobre-pour-la-catégorie-1-2.2">Suppression du mois d'octobre pour la catégorie 1</a></span></li><li><span><a href="#Suppression-du-mois-d'octobre-pour-toutes-les-catégories" data-toc-modified-id="Suppression-du-mois-d'octobre-pour-toutes-les-catégories-2.3">Suppression du mois d'octobre pour toutes les catégories</a></span></li><li><span><a href="#Suppression-des-grands-acheteurs-pour-data_1" data-toc-modified-id="Suppression-des-grands-acheteurs-pour-data_1-2.4">Suppression des grands acheteurs pour data_1</a></span></li><li><span><a href="#Suppression-des-grands-acheteurs-pour-data_2" data-toc-modified-id="Suppression-des-grands-acheteurs-pour-data_2-2.5">Suppression des grands acheteurs pour data_2</a></span></li><li><span><a href="#Grands-acheurs-a-pour-data_1" data-toc-modified-id="Grands-acheurs-a-pour-data_1-2.6">Grands acheurs a pour data_1</a></span></li><li><span><a href="#Grands-acheurs-a-pour-data_2" data-toc-modified-id="Grands-acheurs-a-pour-data_2-2.7">Grands acheurs a pour data_2</a></span></li></ul></li><li><span><a href="#Sauvegade" data-toc-modified-id="Sauvegade-3">Sauvegade</a></span></li></ul></div>

# Projet 4 : Analysez les ventes de votre entreprise

## Analyse globale des données

In [1]:
import pandas as pd
import numpy as np

In [89]:
df_customers = pd.read_csv('dataset_P4\\customers.csv')
df_products = pd.read_csv('dataset_P4\\products.csv')
df_transactions = pd.read_csv('dataset_P4\\transactions.csv')

## 1. df_customers

### Apperçu de la population

In [93]:
df_customers.head()

client_id sex  birth
0    c_4410   f   1967
1    c_7839   f   1975
2    c_1699   f   1984
3    c_5961   f   1962
4    c_5320   m   1943

### Déscription générale des données quantitatifs 

In [4]:
df_customers.describe()

birth
count  8623.000000
mean   1978.280877
std      16.919535
min    1929.000000
25%    1966.000000
50%    1979.000000
75%    1992.000000
max    2004.000000

Dans une vue génrérale, ces statistiques paraissent cohérents

### Vérification des valeurs nulles

In [5]:
df_customers.isnull().sum()

client_id    0
sex          0
birth        0
dtype: int64

### Vérification des erreurs lexicales

In [6]:
df_customers['sex'].value_counts()
# Ou bien
#df_customers.loc[~df_customers['sex'].isin(['m','f'])]

f    4491
m    4132
Name: sex, dtype: int64

### Vérification des individus dupliqués

In [7]:
df_customers.loc[df_customers['client_id'].duplicated(keep=False),:]
# Ou bien
# df_customers['client_id'].value_counts()

Empty DataFrame
Columns: [client_id, sex, birth]
Index: []

Aucune modification n'a été apporté à df_customers

## 2. df_products

### Apperçu de la population

In [8]:
df_products.head()

id_prod  price  categ
0  0_1421  19.99      0
1  0_1368   5.13      0
2   0_731  17.99      0
3   1_587   4.99      1
4  0_1507   3.99      0

### Déscription générale des données quantitatifs 

In [9]:
df_products.describe()
# df_products.describe(include = 'all')

price        categ
count  3287.000000  3287.000000
mean     21.856641     0.370246
std      29.847908     0.615387
min      -1.000000     0.000000
25%       6.990000     0.000000
50%      13.060000     0.000000
75%      22.990000     1.000000
max     300.000000     2.000000

### Détail du produit avec prix négatif 

In [10]:
df_products.loc[df_products['price']<0,:]

id_prod  price  categ
731     T_0   -1.0      0

### Vérification des valeurs nulles

In [11]:
df_products.isnull().sum()

id_prod    0
price      0
categ      0
dtype: int64

### Vérification des erreurs lexicales

In [12]:
# df_products.loc[~df_products['categ'].isin([0,1,2]),:]
df_products['categ'].value_counts()

0    2309
1     739
2     239
Name: categ, dtype: int64

Dans df_products, il y un produit avec prix négatif,</br>

Avant de supprimer cette donnée, je vais viir si elle est utilisé dans df_transactions

## 3. df_transactions

### Apperçu de la population

In [13]:
df_transactions.head()

id_prod                        date session_id client_id
0  0_1483  2021-04-10 18:37:28.723910    s_18746    c_4450
1   2_226  2022-02-03 01:55:53.276402   s_159142     c_277
2   1_374  2021-09-23 15:13:46.938559    s_94290    c_4270
3  0_2186  2021-10-17 03:27:18.783634   s_105936    c_4597
4  0_1351  2021-07-17 20:34:25.800563    s_63642    c_1242

### Déscription générale des données quantitatifs 

In [14]:
df_transactions.describe()

id_prod                             date session_id client_id
count   337016                           337016     337016    337016
unique    3266                           336855     169195      8602
top      1_369  test_2021-03-01 02:30:02.237413        s_0    c_1609
freq      1081                               13        200     12855

### Détail des dates 'test'

In [15]:
masque_dates_test = df_transactions['date'].str.contains('test')
df_transactions[masque_dates_test]

id_prod                             date session_id client_id
1431       T_0  test_2021-03-01 02:30:02.237420        s_0      ct_1
2365       T_0  test_2021-03-01 02:30:02.237446        s_0      ct_1
2895       T_0  test_2021-03-01 02:30:02.237414        s_0      ct_1
5955       T_0  test_2021-03-01 02:30:02.237441        s_0      ct_0
7283       T_0  test_2021-03-01 02:30:02.237434        s_0      ct_1
...        ...                              ...        ...       ...
332594     T_0  test_2021-03-01 02:30:02.237445        s_0      ct_0
332705     T_0  test_2021-03-01 02:30:02.237423        s_0      ct_1
332730     T_0  test_2021-03-01 02:30:02.237421        s_0      ct_1
333442     T_0  test_2021-03-01 02:30:02.237431        s_0      ct_1
335279     T_0  test_2021-03-01 02:30:02.237430        s_0      ct_0

[200 rows x 4 columns]

In [16]:
df_transactions[masque_dates_test].describe()

id_prod                             date session_id client_id
count      200                              200        200       200
unique       1                               39          1         2
top        T_0  test_2021-03-01 02:30:02.237413        s_0      ct_0
freq       200                               13        200       106

### Clients crées pour les dates 'test'

In [17]:
# Dans les dates 'test', il y a 1 seul id_product, 1 seule session_id et 2 client_id 
df_transactions.loc[masque_dates_test,'client_id'].value_counts()

ct_0    106
ct_1     94
Name: client_id, dtype: int64

### Vérification des valeurs manquantes

In [18]:
df_transactions.isnull().sum()

id_prod       0
date          0
session_id    0
client_id     0
dtype: int64

## 4. Relation entre les 3 DF

### Vérifier que tous les 'client_id' de transaction appartiennent à ceux de customers

In [19]:
df_transactions[~df_transactions['client_id'].isin(df_customers['client_id'])]

Empty DataFrame
Columns: [id_prod, date, session_id, client_id]
Index: []

### Vérifier que tous les 'id_product' de transaction appartiennent à ceux de products

In [20]:
df_transactions[~df_transactions['id_prod'].isin(df_products['id_prod'])]

id_prod                        date session_id client_id
6235    0_2245  2021-06-17 03:03:12.668129    s_49705    c_1533
10802   0_2245  2021-06-16 05:53:01.627491    s_49323    c_7954
14051   0_2245  2021-11-24 17:35:59.911427   s_124474    c_5120
17486   0_2245  2022-02-28 18:08:49.875709   s_172304    c_4964
21078   0_2245  2021-03-01 00:09:29.301897        s_3     c_580
...        ...                         ...        ...       ...
322710  0_2245  2021-04-06 19:59:19.462288    s_16936    c_4167
329417  0_2245  2021-03-30 23:29:02.347672    s_13738    c_7790
330490  0_2245  2021-12-03 14:14:40.444177   s_128815    c_6189
335531  0_2245  2021-04-27 18:58:47.703374    s_26624    c_1595
336220  0_2245  2021-05-01 03:35:03.146305    s_28235    c_5714

[103 rows x 4 columns]

In [21]:
df_transactions.loc[~df_transactions['id_prod'].isin(df_products['id_prod']),'id_prod'].value_counts()

0_2245    103
Name: id_prod, dtype: int64

Le produit 0_2245 n'existe pas alors qu'il y a eu 103 transactions!!!!

## Nettoyage des données

In [22]:
masque_dates_test = df_transactions['date'].str.contains('test')
df_transactions[masque_dates_test]

id_prod                             date session_id client_id
1431       T_0  test_2021-03-01 02:30:02.237420        s_0      ct_1
2365       T_0  test_2021-03-01 02:30:02.237446        s_0      ct_1
2895       T_0  test_2021-03-01 02:30:02.237414        s_0      ct_1
5955       T_0  test_2021-03-01 02:30:02.237441        s_0      ct_0
7283       T_0  test_2021-03-01 02:30:02.237434        s_0      ct_1
...        ...                              ...        ...       ...
332594     T_0  test_2021-03-01 02:30:02.237445        s_0      ct_0
332705     T_0  test_2021-03-01 02:30:02.237423        s_0      ct_1
332730     T_0  test_2021-03-01 02:30:02.237421        s_0      ct_1
333442     T_0  test_2021-03-01 02:30:02.237431        s_0      ct_1
335279     T_0  test_2021-03-01 02:30:02.237430        s_0      ct_0

[200 rows x 4 columns]

### Supprimer les dates "test"

In [23]:
df_transactions = df_transactions.loc[~masque_dates_test,:]

### Vérifier si l'un des idendifiants assiciés aux test existe dans les autres bd

In [24]:
# Juste pour vérifier que tout a bien été supprimé 
df_transactions[(df_transactions['id_prod']=='T_0') | (df_transactions['session_id']=='s_0') |
               (df_transactions['client_id']=='ct_0') | (df_transactions['client_id']=='ct_1')]

Empty DataFrame
Columns: [id_prod, date, session_id, client_id]
Index: []

In [25]:
masque_test_prod = df_products['id_prod']=='T_0'
df_products[masque_test_prod]

id_prod  price  categ
731     T_0   -1.0      0

In [26]:
masque_test_cust = (df_customers['client_id']=='ct_0') | (df_customers['client_id']=='ct_1')
df_customers[masque_test_cust]

client_id sex  birth
2735      ct_0   f   2001
8494      ct_1   m   2001

### Suppression des clients et produits relatifs aux dates test 

In [27]:
df_products = df_products.loc[~masque_test_prod,:]
df_customers = df_customers.loc[~masque_test_cust,:]

### Rajout du produit  0_2245

In [90]:
df_products.sort_values(by  = ['categ','id_prod'])

id_prod   price  categ
1001     0_0    3.75      0
803      0_1   10.99      0
2691    0_10   17.95      0
663    0_100   20.60      0
922   0_1000    6.84      0
...      ...     ...    ...
966     2_95   98.99      2
2576    2_96   47.91      2
2698    2_97  160.99      2
3088    2_98  149.74      2
3188    2_99   84.99      2

[3287 rows x 3 columns]

In [91]:
df_products[(df_products['id_prod']>='0_2240') & (df_products['id_prod']<='0_2250') ].sort_values(by  = ['id_prod'])

id_prod  price  categ
2552  0_2240   3.67      0
2482  0_2241  10.99      0
1303  0_2242  31.79      0
2815  0_2243   8.99      0
1725  0_2244   8.37      0
388   0_2246   6.30      0
778   0_2247   2.99      0
3211  0_2248  19.58      0
331   0_2249   2.99      0
2541   0_225  16.99      0
1367  0_2250  20.77      0

In [30]:
# Prix = moyenne des prix de categ 0
moyennePrixCateg0 = round(df_products[df_products['categ']== 0]['price'].mean(),2)

In [31]:
# Rajout d'un enregistrement 
df_products = df_products.append({'id_prod': '0_2245', 'price': moyennePrixCateg0, 'categ' : 0},ignore_index=True)

In [32]:
df_products[df_products['id_prod']=='0_2245']

id_prod  price  categ
3286  0_2245  11.73      0

# Jointure des données

In [33]:
# Jointure entre les 3 DF, 
print(df_transactions.shape)
df_transaction_join = df_transactions.merge(df_products,how='left', on = ['id_prod'])
print(df_transaction_join.shape)
df_transaction_join = df_transaction_join.merge(df_customers, how = 'left', on = ['client_id'])
print(df_transaction_join.shape)
df_transaction_join.head(3)

(336816, 4)
(336816, 6)
(336816, 8)


id_prod                        date session_id client_id  price  categ sex  \
0  0_1483  2021-04-10 18:37:28.723910    s_18746    c_4450   4.99      0   f   
1   2_226  2022-02-03 01:55:53.276402   s_159142     c_277  65.75      2   f   
2   1_374  2021-09-23 15:13:46.938559    s_94290    c_4270  10.71      1   f   

   birth  
0   1977  
1   2000  
2   1979

### Formatage des dates

In [34]:
df_transaction_join['date'] = pd.to_datetime(df_transaction_join['date'], errors='coerce')

In [35]:
df_transaction_join.isnull().sum()


id_prod       0
date          0
session_id    0
client_id     0
price         0
categ         0
sex           0
birth         0
dtype: int64

In [36]:
# df_transaction_join.set_index('date',inplace=True)
# df_transaction_join = df_transaction_join.sort_index()

In [37]:
df_transactions['date'] = pd.to_datetime(df_transactions['date'], errors='coerce')

In [38]:
df_transactions.isnull().sum()


id_prod       0
date          0
session_id    0
client_id     0
dtype: int64

## Rajout des colonnes
### age

In [39]:
df_transaction_join['age'] = 2022-df_transaction_join['birth']

In [40]:
df_transaction_join

id_prod                       date session_id client_id  price  categ  \
0       0_1483 2021-04-10 18:37:28.723910    s_18746    c_4450   4.99      0   
1        2_226 2022-02-03 01:55:53.276402   s_159142     c_277  65.75      2   
2        1_374 2021-09-23 15:13:46.938559    s_94290    c_4270  10.71      1   
3       0_2186 2021-10-17 03:27:18.783634   s_105936    c_4597   4.20      0   
4       0_1351 2021-07-17 20:34:25.800563    s_63642    c_1242   8.99      0   
...        ...                        ...        ...       ...    ...    ...   
336811   1_671 2021-05-28 12:35:46.214839    s_40720    c_3454  31.99      1   
336812   0_759 2021-06-19 00:19:23.917703    s_50568    c_6268  22.99      0   
336813  0_1256 2021-03-16 17:31:59.442007     s_7219    c_4137  11.03      0   
336814   2_227 2021-10-30 16:50:15.997750   s_112349       c_5  50.99      2   
336815  0_1417 2021-06-26 14:38:19.732946    s_54117    c_6714  17.99      0   

       sex  birth  age  
0        f   1977   45  
1        f   2000   22  
2        f   1979   43  
3        m   1963   59  
4        f   1980   42  
...     ..    ...  ...  
336811   m   1969   53  
336812   m   1991   31  
336813   f   1968   54  
336814   f   1994   28  
336815   f   1968   54  

[336816 rows x 9 columns]

### age_range et  age_range_mean

In [41]:
# age_range et  age_range_mean
df_transaction_join['age_range']= pd.cut(df_transaction_join['age'],5)

In [42]:
# age_range_mean
df_transaction_join['age_range_mean']= pd.cut(df_transaction_join['age'],5,labels=[25, 40, 55, 70, 85])

### month

In [43]:
df_transaction_join['month'] = df_transaction_join['date'].dt.month

### price_purchase

In [44]:
df_tmp = df_transaction_join.groupby('client_id')['price'].sum().reset_index(name = 'price_customer_purchase')
df_transaction_join = df_transaction_join.merge(df_tmp, how = 'left', on = 'client_id')
df_transaction_join.head()

id_prod                       date session_id client_id  price  categ sex  \
0  0_1483 2021-04-10 18:37:28.723910    s_18746    c_4450   4.99      0   f   
1   2_226 2022-02-03 01:55:53.276402   s_159142     c_277  65.75      2   f   
2   1_374 2021-09-23 15:13:46.938559    s_94290    c_4270  10.71      1   f   
3  0_2186 2021-10-17 03:27:18.783634   s_105936    c_4597   4.20      0   m   
4  0_1351 2021-07-17 20:34:25.800563    s_63642    c_1242   8.99      0   f   

   birth  age       age_range age_range_mean  month  price_customer_purchase  
0   1977   45    (33.0, 48.0]             40      4                  1186.65  
1   2000   22  (17.925, 33.0]             25      2                   538.48  
2   1979   43    (33.0, 48.0]             40      9                   878.23  
3   1963   59    (48.0, 63.0]             55     10                    33.99  
4   1980   42    (33.0, 48.0]             40      7                   388.52

### price_range

In [45]:
df_transaction_join['price_range']= pd.qcut(df_transaction_join['price'],10).copy()

### number_purchase_month_age, number_purchase_age, purchase_frequency_month et mean_purchase_age

In [46]:
# # number_purchase_month_age
# freq_achat = df_transaction_join.groupby(['age_range_mean','month'])['session_id'].count().reset_index(name = 'number_purchase_month_age')

# # number_purchase_age 
# freq_achat_total = freq_achat.groupby('month')['number_purchase_month_age'].sum().reset_index(name='number_purchase_age')
# freq_achat = freq_achat.merge(freq_achat_total, how='left', on=['month'])

# # purchase_frequency_month
# freq_achat['purchase_frequency_month'] = freq_achat['number_purchase_month_age'] / freq_achat['number_purchase_age']

# # mean_purchase_age
# freq_achat_moyen = freq_achat.groupby('age_range_mean')['number_purchase_month_age'].mean().reset_index(name = 'mean_purchase_age')
# freq_achat = freq_achat.merge(freq_achat_moyen, how='left', on=['age_range_mean'])
# df_transaction_join = df_transaction_join.merge(freq_achat, how='left', on=['age_range_mean','month'])

# # freq_achat =df_transaction_join.groupby([(df_transaction_join.index.year),(df_transaction_join.index.month),'price_range']).count()


### number_purchase

In [67]:
# number_purchase
# df_tmp = df_transaction_join.groupby(['client_id'])['session_id'].count().reset_index(name = 'number_purchase')
# df_transaction_join = df_transaction_join.merge(df_tmp, how='left', on=['client_id'])

#purchase_frequency
# freq_achat = df_tmp['number_purchase'].sum()


# df_tmp['purchase_frequency'] = df_tmp['number_purchase']/freq_achat
# df_transaction_join = df_transaction_join.merge(df_tmp, how='left', on=['client_id'])

# freq_achat =df_transaction_join.groupby([(df_transaction_join.index.year),(df_transaction_join.index.month),'price_range']).count()


In [49]:
# number_purchase_month
df_tmp = df_transaction_join.groupby(['client_id','month'])['session_id'].count().reset_index(name = 'number_purchase_month')
df_transaction_join = df_transaction_join.merge(df_tmp, how='left', on=['client_id','month'])
df_tmp

client_id  month  number_purchase_month
0           c_1      2                      1
1           c_1      6                      1
2           c_1      7                      4
3           c_1      8                      1
4           c_1      9                      2
...         ...    ...                    ...
68029     c_999      8                      3
68030     c_999      9                      5
68031     c_999     10                      2
68032     c_999     11                      5
68033     c_999     12                      1

[68034 rows x 3 columns]

### basket_size

In [50]:
df_panier = df_transaction_join.groupby(['session_id'])['id_prod'].count().reset_index(name = 'basket_size')
df_transaction_join = df_transaction_join.merge(df_panier, how='left', on='session_id')

In [73]:
df_transaction_join['basket_size'].unique()

array([ 3,  2,  9,  1,  7,  5,  4,  6,  8, 10, 11, 13, 14, 12],
      dtype=int64)

### basket_size_range

In [74]:
# basket_size_range
df_transaction_join['basket_size_range']= pd.cut(df_transaction_join['basket_size'],3,labels=['petit panier','panier moyen','grand panier'])

## Suppression du mois d'octobre pour la catégorie 1

In [76]:
data_1 = df_transaction_join[(df_transaction_join["categ"]!=1) | (df_transaction_join['month']!=10)].copy()

## Suppression du mois d'octobre pour toutes les catégories

In [77]:
data_2 = df_transaction_join[(df_transaction_join['month']!=10)].copy()

## Suppression des grands acheteurs pour data_1

In [78]:
q1 = data_1['price_customer_purchase'].quantile(0.25)
q3 = data_1['price_customer_purchase'].quantile(0.75)
iqr = q3 - q1

upper_limit = q3 + 1.5 * iqr
lower_limit = q1 - 1.5 * iqr

In [79]:
data_1[data_1['price_customer_purchase'] > upper_limit]["price_customer_purchase"].unique()

array([162007.34,  73220.8 ,  54466.38, 144257.21])

In [80]:
data_1[data_1['price_customer_purchase'] < lower_limit]["price_customer_purchase"].unique()

array([], dtype=float64)

In [81]:
data_1a = data_1[data_1['price_customer_purchase'] < upper_limit].copy()

## Suppression des grands acheteurs pour data_2

In [82]:
data_2b = data_2[data_2['price_customer_purchase'] < upper_limit].copy()

## Grands acheurs a pour data_1

In [83]:
grands_acheteurs_1 = data_1[data_1['price_customer_purchase']>=20000].copy()

## Grands acheurs a pour data_2

In [84]:
grands_acheteurs_2 = data_2[data_2['price_customer_purchase']>=20000].copy()

# Sauvegade

In [61]:
df_products.to_csv('new_data\\df_products.csv',index=False)
df_customers.to_csv('new_data\\df_customers.csv',index=False)
df_transactions.to_csv('new_data\\df_transactions.csv',index=False)

In [85]:
data_1.to_csv('new_data\\data_1.csv',index=False)
data_2.to_csv('new_data\\data_2.csv',index=False)

In [86]:
data_1a.to_csv('new_data\\data_1a.csv',index=False)
data_2b.to_csv('new_data\\data_2b.csv',index=False)

In [64]:
grands_acheteurs_1.to_csv('new_data\\grands_acheteurs_1.csv',index=False)
grands_acheteurs_2.to_csv('new_data\\grands_acheteurs_2.csv',index=False)

In [65]:
df_transaction_join.to_csv('new_data\\df_transaction_join.csv',index=False)

In [66]:
import matplotlib.pyplot as plt 

In [87]:
df_transaction_join

id_prod                       date session_id client_id  price  categ  \
0       0_1483 2021-04-10 18:37:28.723910    s_18746    c_4450   4.99      0   
1        2_226 2022-02-03 01:55:53.276402   s_159142     c_277  65.75      2   
2        1_374 2021-09-23 15:13:46.938559    s_94290    c_4270  10.71      1   
3       0_2186 2021-10-17 03:27:18.783634   s_105936    c_4597   4.20      0   
4       0_1351 2021-07-17 20:34:25.800563    s_63642    c_1242   8.99      0   
...        ...                        ...        ...       ...    ...    ...   
336811   1_671 2021-05-28 12:35:46.214839    s_40720    c_3454  31.99      1   
336812   0_759 2021-06-19 00:19:23.917703    s_50568    c_6268  22.99      0   
336813  0_1256 2021-03-16 17:31:59.442007     s_7219    c_4137  11.03      0   
336814   2_227 2021-10-30 16:50:15.997750   s_112349       c_5  50.99      2   
336815  0_1417 2021-06-26 14:38:19.732946    s_54117    c_6714  17.99      0   

       sex  birth  age       age_range age_range_mean  month  \
0        f   1977   45    (33.0, 48.0]             40      4   
1        f   2000   22  (17.925, 33.0]             25      2   
2        f   1979   43    (33.0, 48.0]             40      9   
3        m   1963   59    (48.0, 63.0]             55     10   
4        f   1980   42    (33.0, 48.0]             40      7   
...     ..    ...  ...             ...            ...    ...   
336811   m   1969   53    (48.0, 63.0]             55      5   
336812   m   1991   31  (17.925, 33.0]             25      6   
336813   f   1968   54    (48.0, 63.0]             55      3   
336814   f   1994   28  (17.925, 33.0]             25     10   
336815   f   1968   54    (48.0, 63.0]             55      6   

        price_customer_purchase     price_range  number_purchase  \
0                       1186.65   (0.619, 4.99]              100   
1                        538.48  (27.52, 300.0]               15   
2                        878.23   (9.87, 11.73]               64   
3                         33.99   (0.619, 4.99]                3   
4                        388.52    (7.41, 9.87]               28   
...                         ...             ...              ...   
336811                 54466.38  (27.52, 300.0]             3275   
336812                  1168.22  (20.99, 27.52]               95   
336813                  1189.51   (9.87, 11.73]               72   
336814                  1282.37  (27.52, 300.0]               45   
336815                 73220.80  (15.99, 17.99]             4473   

        number_purchase_month  basket_size basket_size_range  
0                           4            3      petit panier  
1                           3            2      petit panier  
2                          20            9      panier moyen  
3                           2            1      petit panier  
4                          10            2      petit panier  
...                       ...          ...               ...  
336811                    276            1      petit panier  
336812                      5            2      petit panier  
336813                      6            1      petit panier  
336814                      3            1      petit panier  
336815                    387            4      petit panier  

[336816 rows x 18 columns]